enclave-runner/src/usercalls/mod.rs

```diff
@@ -346,36 +381,221 @@ impl EnclaveState {
+                    }
+                },
+                CoResult::Return((tcs,v1, v2)) => {
+                    match state.mode {
+                        EnclaveEntry::Library => {
+                            let lib = enclave.kind.as_library().unwrap()
+                                .thread_sender.lock().unwrap();
+                            lib.send(
+                                StoppedTcs {
+                                    tcs,
+                                    event_queue: state.event_queue
+                                }).expect("Work sender couldn't send data to receiver");
+                            Ok((v1, v2))
+                        }
+                        EnclaveEntry::ExecutableMain => {
+                            return Err(EnclaveAbort::MainReturned);
+                        }
+                        EnclaveEntry::ExecutableNonMain => {
+                            assert_eq!(
+                                (v1, v2),
+                                (0, 0),
+                                "Expected enclave thread entrypoint to return zero"
+                            );
+                            let cmd = enclave.kind.as_command().unwrap();
+                            let cmddata = cmd.data.lock().unwrap();
+                            // If the enclave is in the exit-state, threads are no
+                            // longer able to be launched
+                            if !enclave.exiting.load(Ordering::SeqCst) {
+                                cmddata.threads_queue.push(StoppedTcs {
+                                    tcs,
+                                    event_queue: state.event_queue,
+                                });
+                            }
+                            Ok((0,0))
+                        }
+                    }
+                }
+            };
+
+            match (my_result, state.mode ) {
+                (e @ Err(_), EnclaveEntry::Library ) |
+                (e @ Err(_), EnclaveEntry::ExecutableMain ) |
+                (e @ Ok(_), EnclaveEntry::ExecutableMain)=> {
+                    return e;
+                },
+                (Ok(_), EnclaveEntry::ExecutableNonMain) => {
+                    continue;
+                },
+                (Ok(ret), EnclaveEntry::Library) => {
+                    return Ok(ret);
+                }
+                (Err(e @ EnclaveAbort::Exit {..}), EnclaveEntry::ExecutableNonMain) |
+                (Err(e @ EnclaveAbort::InvalidUsercall(_)), EnclaveEntry::ExecutableNonMain) => {
+                    let cmd = enclave.kind.as_command().unwrap();
+                    let mut cmddata = cmd.data.lock().unwrap();
+
+                    if cmddata.primary_panic_reason.is_none() {
+                        cmddata.primary_panic_reason = Some(e)
+                    } else {
+                        cmddata.other_reasons.push(e)
+                    }
+                    return Err(EnclaveAbort::Secondary);
```

Comment:
Why are we returning here, but not if the original error was already `EnclaveAbort::Secondary`?